contracts/WrapperRegistry.sol

```diff
@@ -0,0 +1,110 @@
+     */
+    mapping(address => address) private _wrapperMapping;
+    /**
+     * @dev Enumerable set of underlying token addresses
+     * @dev Required since EnumerableMap only supports uint256->address maps
+     */
+    EnumerableSet.AddressSet private _underlyingTokens;
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Stores address of underlyingToken into `_underlyingTokens`
+     * @dev Stores an entry of underlyingToken address to wrapperToken address into `_wrapperMapping`
+     */
+    function addWrapper(address underlyingToken, address wrapperToken)
+        external
+        override
+        onlyOwner
+        returns (bool)
+    {
+        if (_underlyingTokens.add(underlyingToken)) {
+            _wrapperMapping[underlyingToken] = wrapperToken;
+            emit WrapperAdded(underlyingToken, wrapperToken);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Removes address of underlyingToken from `_underlyingTokens`
+     * @dev Removes the entry of underlyingToken address to wrapperToken address from `_wrapperMapping`
+     */
+    function removeWrapper(address underlyingToken) external override onlyOwner returns (bool) {
+        if (_underlyingTokens.remove(underlyingToken)) {
+            emit WrapperRemoved(underlyingToken, _wrapperMapping[underlyingToken]);
+            delete _wrapperMapping[underlyingToken];
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the length of `_underlyingTokens`
+     */
+    function numWrappers() public view override returns (uint256) {
+        return _underlyingTokens.length();
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev No guarantees are made on the ordering.
+     * @dev Retrieves the underlying token at `index` and returns it with matching wrapper token
+     */
+    function wrapperAt(uint256 index) public view override returns (address, address) {
+        address underlyingToken = _underlyingTokens.at(index);
+        return (underlyingToken, _wrapperMapping[underlyingToken]);
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the wrapper from `_wrapperMapping` by using `underlyingToken` as the key
+     */
+    function getWrapperFromUnderlying(address underlyingToken)
+        external
+        view
+        override
+        returns (address)
+    {
+        return _wrapperMapping[underlyingToken];
+    }
+
+    /**
+     * @inheritdoc IWrapperRegistry
+     * @dev Retrieves the underlying token by iterating over `underlyingTokens`.
+     * @dev Finds first instance that maps to `wrapperToken` in `_wrapperMapping`
+     */
+    function getUnderlyingFromWrapper(address wrapperToken)
+        external
+        view
+        override
+        returns (address)
+    {
+        for (uint256 i = 0; i < numWrappers(); i++) {
```

Comment:
You could just invoke `IWrapper(wrapper).underlying()`? (which is one external storage read, rather than n internal reads)

Reply:
I guess this function also had the purpose of validating that the underlying<->wrapper entry was present in the registry. Since `IWrapper` already has that functionality, it might actually be redundant to include this function at all. Thoughts?

Reply:
I'd elect to remove. You could determine if the wrapper was already in the registry with a staticcall to addWrapper